File: examples/publications/2026/ATVA2026Spectral/Swift/cantor.swift
// Transform with the mem2reg-option
// Verify with VerCors-flag --pallas-sroa

/*@
pure;
requires true;
@*/
func triangular(n: Int) -> Int {
    return (n * (n + 1)) / 2
}

/*@
pure;
ensures _result() == n * n;
@*/
func square(n: Int) -> Int {
    return n * n
}

/*@
pure;
ensures y == 0 ==> _result() == triangular(n: x);
@*/
func cantorPair(x: Int, y: Int) -> Int {
    return (square(n: x + y) + x + (3 * y)) / 2
}

File: examples/publications/2026/ATVA2026Spectral/C/cantor.c
// Transform with mem2reg-option

/*@
declare DEF_RESULT(int);
@*/

/*@
pure;
requires true;
@*/
int triangular(int n) {
    return (n * (n + 1)) / 2;
}

/*@
pure;
ensures _result(int) == n * n;
@*/
int square(int n) {
    return n * n;
}

/*@
pure;
ensures _imply(y == 0,
               _result(int) == triangular(x));
@*/
int cantorPair(int x, int y) {
    return (square(x + y) + x + (3 * y)) / 2;
}